**Part IIB**

MLF_GP1_CreditScore.csv file has credit score data in it. The dataset comprises information on 1700 firms from various industries, including 26 financial and accounting metrics. The last column indicates the rating assigned by Moody's, while the second-to-last column specifies whether the assets are of investment grade or not. Rating of Aaa, Aa1, Aa2, Aa3, A1, A2, A3, Baa1, Baa2, or Baa3 is considered as investment grade. 

**Objective**

The objective is to categorize a company's credit rating into 16 categories and determine if the company is an investment grade. To achieve this, we divide the dataset into training and testing sets in an 80:20 ratio.

**Linear regression**

Linear regression is a statistical approach for determining the connection between one or more independent variables and a dependent variable. The purpose of linear regression is to find the optimal line that describes the connection between the dependent and independent variables.

Ridge and Lasso are linear regression regularisation techniques that add a penalty term to the cost function to prevent overfitting. Ridge regression incorporates an L2 penalty term into the cost function, whereas Lasso regression incorporates an L1 penalty term. The penalty term reduces the coefficients to zero, resulting in a simpler and more understandable model.

The **L2 penalty term** in **Ridge** regression is proportional to the square of the magnitude of the coefficients. This results in lower and more uniformly distributed coefficients, which reduces the effect of multicollinearity (the presence of two or more highly correlated independent variables).

The **L1 penalty term** in **Lasso** regression is proportional to the absolute value of the coefficients. This results in a sparse model with many coefficients that are absolutely zero. Because it may automatically reject irrelevant or redundant independent variables, Lasso regression is beneficial for feature selection.


**CODE :**

The below programme uses linear regression with **Ridge and Lasso regularisation** to classify credit scores. It begins by loading and separating the "**MLF_GP1_CreditScore.csv**" file into training and test sets. For a Ridge regression model, it then turns the target variable into binary labels using **LabelEncoder** and standardises the input features using **StandardScaler**. The model is trained on test data with **alpha = 0.985** and predictions are made. The predicted target probabilities are then turned into binary labels with a threshold of **0.5**, and the label encoder is used to convert them back into their original categories values.

The model's performance is evaluated using **confusion matrix and accuracy score**. Then, the code fits a Lasso regression model on the standardized input features and target values with **alpha=0.1 and max_iter=5000**, and predictions are made on the test data. Finally, the predicted values are converted into binary labels using a **threshold of 0.5** and transformed back into their corresponding investment grade labels using the label encoder. The performance of the Lasso model is evaluated using **confusion matrix and accuracy score**.

In [3]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import accuracy_score, confusion_matrix

# loading the CreditScore creditScoreData file
creditScoreData = pd.read_csv("MLF_GP1_CreditScore.csv")

# Dividing whole data into training and test sets
inputFeatures_trainSet_ridge, inputFeatures_testSet_ridge, target_trainSet_ridge, target_testSet_ridge = train_test_split(creditScoreData.iloc[:, :-2], creditScoreData["InvGrd"], test_size=0.2, random_state=50)

# convert the target variable into binary labels using the LabelEncoder function.
lblEncoder = LabelEncoder()
traget_train_ridge_binary = lblEncoder.fit_transform(target_trainSet_ridge)
traget_test_ridge_binary = lblEncoder.transform(target_testSet_ridge)

# standardizes the input features using the StandardScaler function for a ridge regression model
stdScaler = StandardScaler()
inputFeatures_trainSet_scaled = stdScaler.fit_transform(inputFeatures_trainSet_ridge)
inputFeatures_testSet_ridge_scaled = stdScaler.transform(inputFeatures_testSet_ridge)

# create ridge regression model with regularization parameter alpha = 0.985 and trains the model 
modelRidgeReg = Ridge(alpha=0.985)
modelRidgeReg.fit(inputFeatures_trainSet_scaled, traget_train_ridge_binary)

# predicts the target labels for the test data using the trained ridge regression model
targetPredictedRidge = modelRidgeReg.predict(inputFeatures_testSet_ridge_scaled)

# converts the predicted target probabilities into binary labels using a threshold of 0.5
targetPredictedBinary = (targetPredictedRidge >= 0.5).astype(int)

# converts the predicted binary labels back into their original categorical values using the label encoder
targetPredictedInvGrade = lblEncoder.inverse_transform(targetPredictedBinary)
targetTestInvGradeRidge = lblEncoder.inverse_transform(traget_test_ridge_binary)

# Evaluate model performance
conf_matrix = confusion_matrix(traget_test_ridge_binary, targetPredictedBinary)
accScore = accuracy_score(traget_test_ridge_binary, targetPredictedBinary)

# Displaying the Ridge Regression evaluaiton metrics
print("#Ridge Regression evaluaiton metrics#")
print("Confusion matrix of Ridge Regression model:")
print(conf_matrix)
print("Ridge Regression Model Predicted investment grade:")
print(targetPredictedInvGrade)
print("Ridge Regression Model Actual investment grade")
print(targetTestInvGradeRidge)
print(f"Accuracy of Ridge Regression model: {accScore}")

# Fit Lasso regression model on the standardized input features and target values.
regressionLassoModel = Lasso(alpha=0.1, max_iter=5000)
regressionLassoModel.fit(inputFeatures_trainSet_scaled, traget_train_ridge_binary)

# predict the target variable for the test data using the trained Lasso regression model
targetPredictedLasso = regressionLassoModel.predict(inputFeatures_testSet_ridge_scaled)

# predicts the target variable as binary by converting the predicted values >= 0.5 to 1, otherwise 0
targetPredictedBinary = (targetPredictedLasso >= 0.5).astype(int)

#  Transforms predicted binary values to their corresponding investment grade labels and test target to investment grade labels
targetPredictedInvGrade = lblEncoder.inverse_transform(targetPredictedBinary)
targetTestInvGradeRidge = lblEncoder.inverse_transform(traget_test_ridge_binary)

# Evaluate the performance of the model
accuracy_Score = accuracy_score(traget_test_ridge_binary, targetPredictedBinary)
confus_matrix = confusion_matrix(traget_test_ridge_binary, targetPredictedBinary)

# Print the results for Lasso Regression evaluaiton metrics
print("#Lasso Regression Results#")
print("Lasso model Predicted investment grade labels:")
print(targetPredictedInvGrade)
print("Lasso model Actual investment grade labels:")
print(targetTestInvGradeRidge)
print(f"Accuracy of Lasso model: {accuracy_Score}")
print("Confusion matrix of Lasso model:")
print(confus_matrix)


#Ridge Regression evaluaiton metrics#
Confusion matrix of Ridge Regression model:
[[  4  62]
 [  1 273]]
Ridge Regression Model Predicted investment grade:
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1
 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 1]
Ridge Regression Model Actual investment grade
[0 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 1 1 0
 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 

**Ridge Regression evaluaiton metrics**

Confusion matrix of Ridge Regression model:

    [[  4  62 ]

     [  1 273 ]]

Accuracy of Ridge Regression model: 0.8147058823529412


**Lasso Regression evaluaiton metrics**

Confusion matrix of Lasso model:

    [[  0  66]
     [  0 274] ]

Accuracy of Lasso Regression model: 0.8058823529411765


**Logistic regression**:

Logistic regression is a machine learning approach that is commonly used to solve binary classification problems. Based on the values of the input attributes, it estimates the likelihood of an instance belonging to a specific class (such as 0 or 1). **Ridge and Lasso** regularisation are logistic regression techniques intended to prevent overfitting, which occurs when the model is overly complex and performs well on training data but badly on testing data.

**Ridge regularisation** introduces a penalty term into the loss function that is proportional to the square of the coefficient magnitude. This penalty term reduces the coefficient values to zero, lowering the impact of irrelevant characteristics and improving generalisation performance.

The penalty term in **Lasso regularisation** is proportional to the absolute value of the coefficients. This penalty term effectively performs feature selection and improves the model's interpretability by forcing some of the coefficient values to become exactly zero.

To summarise, adopting Ridge or Lasso regularisation in logistic regression helps to reduce **overfitting**, improves the model's generalisation performance, and can improve its interpretability.

**CODE :**

On a credit score dataset, this code does binary classification using **logistic regression** models with **L1 and L2 regularisation**. Using the pandas package, the dataset is loaded from a CSV file. The categorical variable **'Rating'** is one-hot encoded using the scikit-learn library's **OneHotEncoder** function. The encoded category variable is concatenated with the original data to create a new pandas DataFrame. The data is then divided into training and testing sets in an 80/20 ratio using the train_test_split function from the scikit-learn library.

The **LogisticRegression function from the scikit-learn library** is used to generate two logistic regression models with L1 and L2 regularisation, respectively. The training set is used to train both models. The test data is used to extract the target variables for investment grade prediction.

The trained models are used to forecast the test data's target variables. Both models' evaluation metrics (**accuracy, precision, recall, F1 score, and AUC-ROC**) are obtained using scikit-learn metrics routines. Finally, the evaluation metrics for both models are produced.





In [4]:

# loading the CreditScore creditScoreData file
creditScoreData = pd.read_csv("MLF_GP1_CreditScore.csv")


# one-hot encodes the categorical variable "Rating" using the OneHotEncoder function
ohEnc = OneHotEncoder()
encRatings = ohEnc.fit_transform(creditScoreData['Rating'].values.reshape(-1,1)).toarray()

# concatenate the creditScoreData and the one-hot encoded "Rating" variable into a new pandas DataFrame
creditScoreData_encoded = pd.concat([creditScoreData, pd.DataFrame(encRatings)], axis=1)
# Print the first 5 rows of creditScoreData_encoded dataset
print("First five rows of dataset:")
print(creditScoreData_encoded.head())

#  split the dataset creditScoreData_encoded into train and test sets with a 80/20 ratio
train_creditScoreData_encoded, test_creditScoreData_encoded = train_test_split(creditScoreData_encoded, test_size=0.2, random_state=50)

# splits the train and test data into input features and target variables for investment grade prediction
inputFeatures_test = test_creditScoreData_encoded.iloc[:, :-18]
target_test_invstmntGrade = test_creditScoreData_encoded.iloc[:, -2]

inputFeatures_train = train_creditScoreData_encoded.iloc[:, :-18]
target_train_invstmntGrade = train_creditScoreData_encoded.iloc[:, -2]


#Model 2 Logistic regression:
from sklearn.metrics import accuracy_score, recall_score, precision_score, roc_auc_score, f1_score
from sklearn.linear_model import LogisticRegression

# create a logistic regression model with L1 regularization using the solver 'liblinear' and trains the model on the training data
lrmodel_ridge = LogisticRegression(penalty='l1', solver='liblinear', random_state=50)
lrmodel_ridge.fit(inputFeatures_train, target_train_invstmntGrade)

# predicts target labels for the test data using the trained logistic regression model
predictedTargetInvstmntGrade = lrmodel_ridge.predict(inputFeatures_test)

# compute the evaluation metrics (accuracy, precision, recall, F1 score, and AUC-ROC) for the ridge logistic regression model
prc_lrmodel_ridge = precision_score(target_test_invstmntGrade, predictedTargetInvstmntGrade)
accur_lrmodel_ridge = accuracy_score(target_test_invstmntGrade, predictedTargetInvstmntGrade)
recl_lrmodel_ridge = recall_score(target_test_invstmntGrade, predictedTargetInvstmntGrade)
aucRoc_lrmodel_ridge = roc_auc_score(target_test_invstmntGrade, predictedTargetInvstmntGrade)
f1_lrmodel_ridge = f1_score(target_test_invstmntGrade, predictedTargetInvstmntGrade)


# print the evaluation metrics
print("Ridge Logistic Regression:")
print("Precision:", prc_lrmodel_ridge)
print("Accuracy:", accur_lrmodel_ridge)
print("Recall:", recl_lrmodel_ridge)
print("AUC-ROC:", aucRoc_lrmodel_ridge)
print("F1-score:", f1_lrmodel_ridge)


# create a logistic regression model with L2 regularization using the solver 'liblinear' and trains the model on the training data
lrmodel_lasso = LogisticRegression(penalty='l2', solver='liblinear', random_state=50)
lrmodel_lasso.fit(inputFeatures_train, target_train_invstmntGrade)

# make predictions on the test creditScoreData
predictedTargetInvstmntGradeLasso = lrmodel_lasso.predict(inputFeatures_test)

# compute the evaluation metrics (accuracy, precision, recall, F1 score, and AUC-ROC) for the lasso logistic regression model
prc_lrmodel_lasso = precision_score(target_test_invstmntGrade, predictedTargetInvstmntGradeLasso)
accur_lrmodel_lasso = accuracy_score(target_test_invstmntGrade, predictedTargetInvstmntGradeLasso)
aucRoc_lrmodel_lasso = roc_auc_score(target_test_invstmntGrade, predictedTargetInvstmntGradeLasso)
recl_lrmodel_lasso = recall_score(target_test_invstmntGrade, predictedTargetInvstmntGradeLasso)
f1_lrmodel_lasso = f1_score(target_test_invstmntGrade, predictedTargetInvstmntGradeLasso)


# print the evaluation metrics
print("Lasso Logistic Regression:")
print("Precision:", prc_lrmodel_lasso)
print("Accuracy:", accur_lrmodel_lasso)
print("AUC-ROC:", aucRoc_lrmodel_lasso)
print("Recall:", recl_lrmodel_lasso)
print("F1-score:", f1_lrmodel_lasso)


First five rows of dataset:
   Sales/Revenues  Gross Margin    EBITDA  EBITDA Margin  \
0       -0.005496      0.030763  0.018885       0.024515   
1       -0.005496      0.030763  0.088716       0.094733   
2       -0.007045      0.023159  0.088716       0.096440   
3       -0.009396      0.028400  0.088716       0.099046   
4       -0.009009      0.027714  0.088716       0.098611   

   Net Income Before Extras  Total Debt  Net Debt   LT Debt   ST Debt  \
0                  0.146849   -0.029710 -0.019296 -0.042648  0.049875   
1                  0.146849   -0.029710 -0.019296 -0.042648  0.049875   
2                  0.108590    0.039410  0.034268  0.009059  0.250371   
3                  0.146137    0.030071  0.036938 -0.016964  0.356994   
4                  0.123500    0.024224  0.034445 -0.034132  0.461894   

       Cash  ...    6    7    8    9   10   11   12   13   14   15  
0 -0.133716  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1 -0.133716  ...  0.0  0.0  0.0  0

**Ridge Logistic Regression Evaluation Metrics**:

Precision: 0.2

Accuracy: 0.8382352941176471

Recall: 0.019230769230769232

AUC-ROC: 0.5026709401709403

F1-score: 0.03508771929824562

The assessment metrics displayed are for the Ridge Logistic Regression model's performance on a binary classification test. Each metric is defined as follows:

- Precision is defined as the ratio of true positive predictions to total positive predictions. In this scenario, it means that just 20% of all anticipated positive instances were real positive. 

- Accuracy: This is the ratio of correct forecasts to total predictions made. In this situation, the model's accuracy is 83.82%, which means that 83.82% of all predictions were true.

- Recall: This is the ratio of true positive predictions to total positive cases in the data. In this scenario, it means that the model correctly identified only 1.92% of all positive instances.

- AUC-ROC. The ROC curve is a graph that compares the true positive rate to the false positive rate at various categorization levels. The AUC-ROC score ranges from 0 to 1, with 0.5 representing random guessing and 1 representing perfect classification. The AUC-ROC score in this situation is 0.502, indicating that the model performs only marginally better than random guessing.

- F1-score: This is the harmonic mean of precision and recall. It achieves a good combination of precision and recall. The F1-score in this situation is 0.035, indicating that the model is underperforming in terms of both precision and recall.


**Lasso Logistic Regression Evaluation Metrics:**

Precision: 0.2

Accuracy: 0.8382352941176471

AUC-ROC: 0.5026709401709403

Recall: 0.019230769230769232

F1-score: 0.03508771929824562

These are the measures for evaluating a Lasso logistic regression model.

Precision is defined as the proportion of genuine positive predictions to total anticipated positives. A accuracy of 0.2 suggests that just 20% of all anticipated positive instances were true positives.

Accuracy is defined as the proportion of correctly classified cases to total instances. The model accurately classified 83.82% of the occurrences with an accuracy of 0.8382.

AUC-ROC stands for Area Under the Receiver Operating Characteristic Curve, and it indicates the model's ability to differentiate between positive and negative classes. An AUC-ROC of 0.5027 indicates that the model's ability to differentiate across classes is comparable to random chance.

Recall is defined as the proportion of genuine positive predictions to total actual positives. A recall of 0.0192 indicates that the model accurately detected just 1.92% of all actual positive events.

F1-score: The harmonic mean of precision and recall is the F1-score. A F1-score of 0.035 indicates that the model has a poor balance of precision and recall.


**Neural network model**:

Building a neural network model that can learn and recognise patterns in the input data and forecast the related output is used in a Neural Networks-based technique to classifying the firm's rating into one of the rating categories and predicting if it is in an investment grade. This problem's input data is the firms' financial data, which includes numerous attributes such as liquidity ratios, profitability ratios, leverage ratios, and so on. The output variable is the firm's rating, which is a categorical variable with numerous possible values (for example, AAA, AA, A, BBB, and so on).

Furthermore, the aim is to forecast if the firm is of investment grade or not, which is a categorical variable with two possible values (yes or no).

Various neural network topologies, such as Feedforward Neural Networks, Convolutional Neural Networks, and Recurrent Neural Networks, can be used to develop a Neural Networks-based strategy. The model architecture is determined by the type of data and the complexity of the task. The model is trained on a labelled dataset that contains input features and associated output labels. The training procedure entails modifying the neural network's weights and biases to minimise the difference between the predicted and actual output. To evaluate the model's performance, it is validated using a different test dataset. Finally, the model can be used to forecast new, previously unknown data.


**CODE**:

The programme uses a **Neural Network-based** approach to categorise the firm's rating and predict if it is investment grade or not.

The method begins by loading the credit score data file and utilising a dictionary to translate the rating categories to investment grade. The data is then divided into training and test sets, and one-hot encoding is applied to the target labels. To centre and normalise the data, the input features are scaled using **StandardScaler**.

Following that, a **Sequential Neural Network** model with two hidden layers, dropout regularisation, and softmax activation is built. The model is built with **categorical_crossentropy** as a loss function, the **Adam** optimizer, and the **accuracy** metric in mind.

The fit approach is used to train the **Sequential Neural Network model** with the training data, and the predict method is used to generate the predicted target labels for the test data. The accuracy_score function from the **scikit-learn** library is used to calculate the trained model's accuracy score.

Finally, the code returns the anticipated and actual investment grade labels to the grading categories and outputs the model's accuracy score.
 

In [5]:

from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Dense, Dropout
from sklearn.metrics import confusion_matrix,accuracy_score
from tensorflow.keras.models import Sequential


# loading the CreditScore data file
creditScoreData = pd.read_csv("MLF_GP1_CreditScore.csv")
print(creditScoreData.shape)

# Mapping ratings to investment grade
ratingToInvestmentMap = {
              'Baa1': 1, 'Baa2': 1, 'Baa3': 1, 'Ba1': 0, 'Ba2': 0, 'Ba3': 0,
			  'Aaa': 1, 'Aa1': 1, 'Aa2': 1, 'Aa3': 1, 'A1': 1, 'A2': 1, 'A3': 1,
              'B1': 0, 'B2': 0, 'B3': 0, 'Caa1': 0}
creditScoreData["InvGrd"] = creditScoreData["Rating"].map(ratingToInvestmentMap)

# Dividing whole data into training and test sets
inputFeatures_train, inputFeatures_test, target_train, target_test = train_test_split(creditScoreData.iloc[:, :-2], creditScoreData["InvGrd"], test_size=0.1, random_state=50)

# The to_categorical function produces a new array in which each element is transformed to a one-hot encoded vector
target_train = to_categorical(target_train) 
target_test = to_categorical(target_test)

# Scale the input features by centering and normalizing the data using the StandardScaler function
stndrdSclr = StandardScaler()
inputFeatures_train_scaled = stndrdSclr.fit_transform(inputFeatures_train)
inputFeatures_test_scaled = stndrdSclr.transform(inputFeatures_test)

# Create sequential neural network model with 2 hidden layers, dropout regularization, and softmax activation
sequentialNNModel = Sequential()
sequentialNNModel.add(Dense(32, input_dim=inputFeatures_train_scaled.shape[1], activation="relu"))
sequentialNNModel.add(Dropout(0.5))
sequentialNNModel.add(Dense(16, activation="relu"))
sequentialNNModel.add(Dropout(0.5))
sequentialNNModel.add(Dense(2, activation="softmax"))
sequentialNNModel.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])


# train the sequential neural network model with scaled input features and target labels.
sequentialNNModel.fit(inputFeatures_train_scaled, target_train, epochs=50, batch_size=32, verbose=1)

# predicts target labels for the test data using the trained sequential neural network model
predicted_target = sequentialNNModel.predict(inputFeatures_test_scaled)
predicted_target_class = predicted_target.argmax(axis=-1)

# map the predicted and actual investment grade labels back to the rating categories
predicted_target_rating = [list(ratingToInvestmentMap.keys())[list(ratingToInvestmentMap.values()).index(i)] for i in predicted_target_class]
target_test_rating = [list(ratingToInvestmentMap.keys())[list(ratingToInvestmentMap.values()).index(j)] for j in target_test.argmax(axis=-1)]

# evaluate the model's performance
SeqNNAccScore = accuracy_score(target_test.argmax(axis=-1), predicted_target_class)


print("Accuracy of the trained sequential neural network model:", SeqNNAccScore)


(1700, 28)
Epoch 1/50
48/48 [==============================] - 1s 2ms/step - loss: 0.9371 - accuracy: 0.4935
Epoch 2/50
48/48 [==============================] - 0s 2ms/step - loss: 0.7880 - accuracy: 0.6771
Epoch 3/50
48/48 [==============================] - 0s 2ms/step - loss: 0.7224 - accuracy: 0.7144
Epoch 4/50
48/48 [==============================] - 0s 2ms/step - loss: 0.6705 - accuracy: 0.7366
Epoch 5/50
48/48 [==============================] - 0s 2ms/step - loss: 0.6371 - accuracy: 0.7471
Epoch 6/50
48/48 [==============================] - 0s 2ms/step - loss: 0.6191 - accuracy: 0.7556
Epoch 7/50
48/48 [==============================] - 0s 2ms/step - loss: 0.6019 - accuracy: 0.7634
Epoch 8/50
48/48 [==============================] - 0s 2ms/step - loss: 0.6091 - accuracy: 0.7523
Epoch 9/50
48/48 [==============================] - 0s 2ms/step - loss: 0.6096 - accuracy: 0.7497
Epoch 10/50
48/48 [==============================] - 0s 2ms/step - loss: 0.5763 - accuracy: 0.7562
Epoch 11

**Accuracy of the trained sequential neural network model: 0.8058823529411765**

The trained sequential neural network model accurately categorised 80.59% of the test data instances, resulting in an accuracy of 0.8058823529411765. In other words, the model accurately classified 80.59% of the test data cases. This is a measure of how well the model predicts whether a company's credit rating is investment grade or not.
